storage: Add a call to reserve a replica before trying to add a replica

BramGruneir opened the pull request:
- storage: remove timeout between rebalance attempts
  This is no longer needed since we have a combination of both preemptive
  snapshots and the reservation system.
- storage: add Reserve RPCs to replica's ChangeReplicas
  This also adds functionality to the store pool to stop repeated attempts to
  reserve a replica on a store if one is declined.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7147)

<!-- Reviewable:end -->

BramGruneir committed:
storage: add Reserve RPC calls to replica's ChangeReplicas

This also adds functionality to the store pool to stop repeated attempts to
reserve a replica on a store if one is declined.

BramGruneir committed:
storage: remove timeout between rebalance attempts

This is no longer needed since we have a combination of both preemptive
snapshots and the reservation system.

tamird commented:
Reviewed 4 of 4 files at r1, 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 149 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrAnIORqcEUe3EXvgL:-KJrAnIP-kIO6xHgp-oD:788013188) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/server/context.go#L149)):*
> ```Go
> 
> 	// ReservationsEnabled is a switch used to enable the add replica
> 	// reservation system
> ```

add a period

---

*[storage/allocator_test.go, line 172 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrB7-7l7blRBA18mJz:-KJrB7-7l7blRBA18mK-:-1293635160) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/allocator_test.go#L172)):*
> ```Go
> 	// Have to call g.SetNodeID before call g.AddInfo
> 	g.SetNodeID(roachpb.NodeID(1))
> 	storePool := NewStorePool(g, clock, rpcContext, false, TestTimeUntilStoreDeadOff, stopper)
> ```

throughout: consider a `const TestReservationsEnabled = false` to aid readability

---

*[storage/replica_command.go, line 2546 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrIOpAeFmyEjmgYh0A:-KJrIOpAeFmyEjmgYh0B:-1662358805) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/replica_command.go#L2546)):*
> ```Go
> 			r.store.StoreID(),
> 			replica.StoreID,
> 			r.RangeID,
> ```

s/r.RangeID/rangeID/

---

*[storage/replica_command.go, line 2549 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrIKjgSRGW_rhW_ACp:-KJrIKjgSRGW_rhW_ACq:-700152489) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/replica_command.go#L2549)):*
> ```Go
> 			maxSize,
> 		); err != nil {
> 			return err
> ```

local convention says `return util.Errorf("change replicas of %d failed: %s", rangeID, err)`

---

*[storage/store_pool.go, line 87 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrBig7WW7SmIuZeSVQ:-KJrBig7WW7SmIuZeSVR:-946629539) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L87)):*
> ```Go
> 	// alive.
> 	if sd.dead || sd.desc == nil {
> 		return
> ```

oh dear. can we avoid naked returns please? also, the store can't match if it isn't alive, right? perhaps an enum is better than two bools here.

---

*[storage/store_pool.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCWOmRb8LPo86E-zz:-KJrCWOmRb8LPo86E0--:603921796) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L399)):*
> ```Go
> // TODO(bram): consider moving the nodeID to the store pool during
> // NewStorePool.
> func (sp *StorePool) requestReservation(
> ```

naming nit: `reserve`? there's no return other than an error.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KJrCifMSFHnt0BodlqC:2004537431) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
> ```Go
> 	}
> 	sp.mu.Lock()
> 	defer sp.mu.Unlock()
> ```

do you need to hold the lock during the RPC?

---

*[storage/store_pool.go, line 441 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCqKQdw_tdRVznEg8:-KJrCqKQdw_tdRVznEg9:-41378749) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L441)):*
> ```Go
> 	if err != nil {
> 		detail.reservationDeclinedOn = sp.clock.Now().GoTime()
> 		return fmt.Errorf("reservation declined:%+v due to error:%s", req, err)
> ```

not technically true that it was declined.

---

*[storage/store_pool_test.go, line 425 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrHK_OEuX3FEJe3dMN:-KJrHK_OEuX3FEJe3dMO:-1529376261) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L425)):*
> ```Go
> 
> func (f fakeNodeServer) Batch(ctx context.Context, args *roachpb.BatchRequest) (*roachpb.BatchResponse, error) {
> 	panic("unimplemented")
> ```

is there any reason for Batch and Reserve to belong to the same grpc Service? I think not, and separating them would be a bit cleaner (and not require much code to change).

---

*[storage/store_pool_test.go, line 432 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrI-aN3taax5jYPPTF:-KJrI-aN3taax5jYPPTG:365464043) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L432)):*
> ```Go
> }
> 
> func (f fakeNodeServer) Reserve(_ context.Context, _ *roachpb.ReservationRequest) (*roachpb.ReservationResponse, error) {
> ```

not a good idea to implement interfaces on a value :(

use a pointer.

---

*[storage/store_pool_test.go, line 438 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrHbrta5aFpJweKC7L:-KJrHbrta5aFpJweKC7M:348853662) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L438)):*
> ```Go
> // newFakeNodeServer returns a newFakeNodeServer designed to handle internal
> // node server RPCs.
> func newFakeNodeServer(t *testing.T, ctx *rpc.Context) (*grpc.Server, net.Listener) {
> ```

i don't see a reason to return this grpc server - you should return the fakeNodeServer instead.

---

*[storage/store_pool_test.go, line 453 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrHpV5xD8rxFZqSmPl:-KJrHpV5xD8rxFZqSmPm:-875357286) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L453)):*
> ```Go
> func newFakeNodeServerContext(clock *hlc.Clock, stopper *stop.Stopper) *rpc.Context {
> 	ctx := rpc.NewContext(testutils.NewNodeTestBaseContext(), clock, stopper)
> 	ctx.HeartbeatInterval = 10 * time.Millisecond
> ```

why are these intervals jacked up?

---

*[storage/store_pool_test.go, line 536 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrIBJLobWBgm-66-Ya:-KJrIBJLobWBgm-66-Yb:-1055684445) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L536)):*
> ```Go
> 			t.Errorf("%d: expected an error:%s, got none", i, testCase.expErr)
> 		} else if !testutils.IsError(err, testCase.expErr) {
> 			t.Errorf("%d: expected error:%s, actual %s", i, testCase.expErr, err)
> ```

s/actual/actual:/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 149 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrAnIORqcEUe3EXvgL:-KK79GBuDMQWp75w0_tX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/server/context.go#L149)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> add a period
</details>
Done.

---

*[storage/allocator_test.go, line 172 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrB7-7l7blRBA18mJz:-KK9SsFrfBJhSmgtXNhr:1866294270) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/allocator_test.go#L172)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: consider a `const TestReservationsEnabled = false` to aid readability
</details>
Instead, I just added a comment into the call everytime.
/* reservationsEnabled */ false,

---

*[storage/replica_command.go, line 2546 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrIOpAeFmyEjmgYh0A:-KK9UIx-M4hvSRyFquUM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/replica_command.go#L2546)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/r.RangeID/rangeID/
</details>
Done.

---

*[storage/replica_command.go, line 2549 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrIKjgSRGW_rhW_ACp:-KK9UQA25q1RHUf7NYf5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/replica_command.go#L2549)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> local convention says `return util.Errorf("change replicas of %d failed: %s", rangeID, err)`
</details>
Done.

---

*[storage/store_pool.go, line 87 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrBig7WW7SmIuZeSVQ:-KK9IWfUraHGMhTr-OKy:12347095) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L87)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oh dear. can we avoid naked returns please? also, the store can't match if it isn't alive, right? perhaps an enum is better than two bools here.
</details>
Excellent idea.  I really didn't like having the two bools returned and the named returns made me squirm as well.  
Done.

---

*[storage/store_pool.go, line 399 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCWOmRb8LPo86E-zz:-KK9IuOydck4vnW109oh:2129167674) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L399)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> naming nit: `reserve`? there's no return other than an error.
</details>
Good call.  Done.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KK9MwMOJn9_RcrHyvmc:-545814060) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> do you need to hold the lock during the RPC?
</details>
nope, fixed and added a check to make sure nothing funky happens when not in the lock.

---

*[storage/store_pool.go, line 441 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCqKQdw_tdRVznEg8:-KK9N8KTUFhQDO0h0sGz:-1372395156) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L441)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not technically true that it was declined.
</details>
replaced with failed

---

*[storage/store_pool_test.go, line 425 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrHK_OEuX3FEJe3dMN:-KK9NE45sI8zqFobAslu:230957882) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L425)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is there any reason for Batch and Reserve to belong to the same grpc Service? I think not, and separating them would be a bit cleaner (and not require much code to change).
</details>
Both pollfrozen and reserve make sense together, both are targeting a store on a node.  I'm not going to change this now, but I'll add an issue to split them. #7189

---

*[storage/store_pool_test.go, line 432 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrI-aN3taax5jYPPTF:-KK9Ngsv6VNetvWqawF7:-1711306107) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L432)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not a good idea to implement interfaces on a value :(
> 
> use a pointer.
</details>
ick, yeah, done.

---

*[storage/store_pool_test.go, line 438 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrHbrta5aFpJweKC7L:-KK9R3b0g7mZIypcSo_G:1777913557) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L438)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i don't see a reason to return this grpc server - you should return the fakeNodeServer instead.
</details>
yeah, no need for it.  Cleaned up the initialization of the fakeNodeServer.

---

*[storage/store_pool_test.go, line 453 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrHpV5xD8rxFZqSmPl:-KK9RO-TEfmvchNeRbn6:1642000329) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L453)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why are these intervals jacked up?
</details>
Was copied from the code in send_test.go.  Don't think they're needed, so removed.

---

*[storage/store_pool_test.go, line 536 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrIBJLobWBgm-66-Ya:-KK9SKRSQKC7JHbYqVcj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool_test.go#L536)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/actual/actual:/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 11 of 11 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KK9vtSxUU-iBFejOFy_:-1961393896) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> nope, fixed and added a check to make sure nothing funky happens when not in the lock.
</details>
oh, i didn't realize the update below was happening under lock - i think you do need this lock to avoid write skew.

---

*[storage/store_pool_test.go, line 442 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KK9wFWt_sP0h0XL0SGR:-KK9wFWt_sP0h0XL0SGS:-1682748590) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e5851db9d71a040b91daea52d1e707b40c5dde/storage/store_pool_test.go#L442)):*
> ```Go
> }
> 
> // newFakeNodeServer returns a newFakeNodeServer designed to handle internal
> ```

a newFakeNodeServer is not a type

---

*[storage/store_pool_test.go, line 445 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KK9wQ29Swv-2j1W0GES:-KK9wQ29Swv-2j1W0GET:-71062932) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e5851db9d71a040b91daea52d1e707b40c5dde/storage/store_pool_test.go#L445)):*
> ```Go
> // node server RPCs, an rpc context used for the server and a listener. The
> // listener should be closed when it is no longer needed.
> func newFakeNodeServer(t *testing.T, stopper *stop.Stopper) (*fakeNodeServer, *rpc.Context, net.Listener) {
> ```

nit: i'd make this return an error rather than take a testing.T

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 13 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KKAVfM2VhFLIAX8s7JA:-691534683) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oh, i didn't realize the update below was happening under lock - i think you do need this lock to avoid write skew.
</details>
I'm not worried about it actually.  We don't ever replace or delete storeDetails, but even if we do, this would only add a 10ish second delay before it's considered for rebalancing again.  This should actually improve performance.  But it would allow more than one range to potentially call the same RPC during that time (but this would only happen if you have more than one store running at a time, so again, not a big problem).  

---

*[storage/store_pool_test.go, line 442 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KK9wFWt_sP0h0XL0SGR:-KKAWb0axIU6t1kQop2u:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e5851db9d71a040b91daea52d1e707b40c5dde/storage/store_pool_test.go#L442)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> a newFakeNodeServer is not a type
</details>
Done.

---

*[storage/store_pool_test.go, line 445 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KK9wQ29Swv-2j1W0GES:-KKAYTBSiTcobsbHr75f:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e5851db9d71a040b91daea52d1e707b40c5dde/storage/store_pool_test.go#L445)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: i'd make this return an error rather than take a testing.T
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 11 of 11 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KKAYu58lKhjT1uDKwhS:-1419087291) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I'm not worried about it actually.  We don't ever replace or delete storeDetails, but even if we do, this would only add a 10ish second delay before it's considered for rebalancing again.  This should actually improve performance.  But it would allow more than one range to potentially call the same RPC during that time (but this would only happen if you have more than one store running at a time, so again, not a big problem).  
</details>
why does it require that storeDetails are deleted or replaced? if you have two calls to this function that race, you get write skew. what am I missing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KKA_BQ7_TbYlw01xxd7:1273554844) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why does it require that storeDetails are deleted or replaced? if you have two calls to this function that race, you get write skew. what am I missing?
</details>
Write skew is fine.  It would set the last time declined at t0, then the other would set it at t1.  If we have a slight difference in times, or even the orders are swapped, it doesn't change how the storeDetail will be handled.  It would just mean a very slightly longer or shorter delay.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KKA_zgYQiTMI_kGtvvF:1704242496) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Write skew is fine.  It would set the last time declined at t0, then the other would set it at t1.  If we have a slight difference in times, or even the orders are swapped, it doesn't change how the storeDetail will be handled.  It would just mean a very slightly longer or shorter delay.
</details>
OK, maybe it's fine today, but isn't this asking for trouble down the road?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KKAbc3Gb3n725oZ64GE:-2100892044) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK, maybe it's fine today, but isn't this asking for trouble down the road?
</details>
I don't see how.   Since we're locking around the write itself, it will never go backwards, which would be the  only weird case.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KKAd8NY4f5U5il9IvRW:-516862143) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I don't see how.   Since we're locking around the write itself, it will never go backwards, which would be the  only weird case.  
</details>
This is classic write skew
```
A read
B read
B rpc
B write
A rpc
A write
```
what happened to B's update?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Glad to see our temporary rebalancer backoff code go away!

<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> #### storage: Add a call to reserve a replica before trying to add a replica
> 
> - storage: remove timeout between rebalance attempts
> 
> This is no longer needed since we have a combination of both preemptive
> 
> snapshots and the reservation system.
> 
> 
> 
> - storage: add Reserve RPCs to replica's ChangeReplicas
> 
> This also adds functionality to the store pool to stop repeated attempts to
> 
> reserve a replica on a store if one is declined.
> 
> 
> 
</details>


---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/replica_command.go, line 2550 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKA_VIV-nWVuM9BMjc7:-KKA_VIV-nWVuM9BMjc8:-1985342600) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/replica_command.go#L2550)):*
> ```Go
> 		}
> 
> 		// Before we try to add a new replica, we first need to secure a
> ```

The second sentence is superfluous, because it's implied from the first.

---

*[storage/store_pool.go, line 445 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKAaHOmNtFYwFPo6O0U:-KKAaHOnKQ_S0tttiYoP:1264240886) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L445)):*
> ```Go
> 		RangeID:     rangeID,
> 	}
> 	resp, err := client.Reserve(context.Background(), req)
> ```

Is there a timeout for this RPC? What happens to replicate queue performance if there's a node that's slow to respond to `Reserve` requests?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-:-KJuXd7YoTgdhp0VXyDd:637780993)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/store_pool.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KKAe4A-WbPCU-bsTSNF:894732886) ([raw file](https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> This is classic write skew
> ```
> A read
> B read
> B rpc
> B write
> A rpc
> A write
> ```
> what happened to B's update?
> 
</details>
Again, in this case, it doesn't matter.  
B's write would be gone when A's write comes along and overrides it.  But it would do so with a slightly later time.

Since the value in question is just a "don't try to call reserve for this long" the error here would actually be B's read.  And if we happen to send 2 or 3 RPCs in quick succession, it won't cause any problems.  Since the replicate queue is the only way in which this gets called in the first place, we'll only get a single call per store anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

tamird commented:
...then why not hold the lock? Your explanation makes sense, but this a
subtlety that is just not needed here, and creates potential for error when
combined with future changes.
On Jun 13, 2016 17:33, "Bram Gruneir" <notifications@github.com> wrote:

> Review status: all files reviewed at latest revision, 3 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *storage/store_pool.go, line 410 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KJrCifMSFHnt0BodlqB:-KKAe4A-WbPCU-bsTSNF:894732886>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/569decc5cfafb074ec728cdbbdcaf024a03ee54d/storage/store_pool.go#L410>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> This is classic write skew
>
> A read
> B read
> B rpc
> B write
> A rpc
> A write
>
> what happened to B's update?
>
>
> Again, in this case, it doesn't matter.
>
> B's write would be gone when A's write comes along and overrides it. But
> it would do so with a slightly later time.
>
> Since the value in question is just a "don't try to call reserve for this
> long" the error here would actually be B's read. And if we happen to send 2
> or 3 RPCs in quick succession, it won't cause any problems. Since the
> replicate queue is the only way in which this gets called in the first
> place, we'll only get a single call per store anyway.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7147>*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7147#issuecomment-225715550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABdsPHhbJ2nqb7hrCk3qq5y_OS6QzP2qks5qLcyqgaJpZM4Iychi>
> .
>

BramGruneir commented:
Sounds good.  I'll add the lock back in.

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ...then why not hold the lock? Your explanation makes sense, but this a subtlety that is just not needed here, and creates potential for error when combined with future changes. 
</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-:-KKAmvwCxrkHOQgbjsCN:-153715377)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 305 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC54C4fZzQb9IX4I-5:-KKC54C4fZzQb9IX4I-6:1074313325) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/server/context.go#L305)):*
> ```Go
> 	ctx.TimeUntilStoreDead = envutil.EnvOrDefaultDuration("time_until_store_dead", ctx.TimeUntilStoreDead)
> 	ctx.ConsistencyCheckInterval = envutil.EnvOrDefaultDuration("consistency_check_interval", ctx.ConsistencyCheckInterval)
> 	ctx.ReservationsEnabled = envutil.EnvOrDefaultBool("reservations_enabled", ctx.ReservationsEnabled)
> ```

Add a TODO to remove this after a short testing period? Removing both the reservation system and the old backoff code doesn't sound like a stable configuration.

---

*[storage/allocator_test.go, line 176 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKBzvD0Au2lCwXZ1f2W:-KKBzvD0Au2lCwXZ1f2X:1045887536) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/allocator_test.go#L176)):*
> ```Go
> 		clock,
> 		rpcContext,
> 		/* reservationsEnabled */ false,
> ```

Do these tests fail with reservations enabled? Since this is going to be the default we should generally be going through this code path in tests. Maybe we need to mock out the RPC itself (and that can be left as a TODO), but we should try to turn this on wherever we can (especially in `multiTestContext`)

---

*[storage/replica_command.go, line 2560 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC0PyDk9P-1hUcBOQ4:-KKC0PyDk9P-1hUcBOQ5:751367439) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/replica_command.go#L2560)):*
> ```Go
> 		if err := r.store.allocator.storePool.reserve(
> 			r.store.nodeDesc.NodeID,
> 			r.store.StoreID(),
> ```

Nit: We should be getting the NodeID and StoreID in the same way here. Maybe just pass `r.store.Ident` into `reserve`?

Why does `reserve` need the node and store ID for the sender but only the store ID for the target?

---

*[storage/store_pool.go, line 101 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC1I5rxS7wy9vv2rng:-KKC1I5rxS7wy9vv2rnh:-1504341004) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L101)):*
> ```Go
> 	// The store must not have a recent declined reservation to be considered
> 	// available for matching.
> 	if sd.reservationDeclinedOn.Add(timeoutForDeclinedReservations).After(now) ||
> ```

It's weird that a throttled target is "alive" instead of either "dead" or a distinct value. In what situation would you want to treat these two cases the same?

---

*[storage/store_pool.go, line 390 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC2Al-9DrBcPequV0o:-KKC2Al-9DrBcPequV0p:-1767175099) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L390)):*
> ```Go
> 		detail := sp.mu.stores[roachpb.StoreID(storeID)]
> 		matched := detail.match(now, required)
> 		if matched >= storeMatchAlive {
> ```

A throttled store should only be counted in `aliveStoreCount` if it matches the required attributes. 

---

*[storage/store_pool.go, line 426 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC2vctz8C9hfM0sRG_:-KKC2vcupc-dG8NIZUlm:83764518) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L426)):*
> ```Go
> 	}
> 	address := detail.desc.Node.Address.String()
> 	nodeID := detail.desc.Node.NodeID
> ```

s/nodeID/toNodeID/

---

*[storage/store_pool_test.go, line 451 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC3ZhW0xPSXM0fHmFK:-KKC3ZhW0xPSXM0fHmFL:1074968438) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool_test.go#L451)):*
> ```Go
> 	if err != nil {
> 		return nil, nil, ln, err
> 	}
> ```

Can you do `stopper.AddCloser(ln)` here so the caller doesn't have to?

---

*[storage/store_pool_test.go, line 507 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC4-WE4n9LhnpKNA9y:-KKC4-WE4n9LhnpKNA9z:1259502286) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool_test.go#L507)):*
> ```Go
> 		fakeErr  string
> 		storeID  int
> 		expErr   string //Regex Pattern
> ```

Space after `//`. Or just remove the comment; doesn't look like the regex usage matters.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 305 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC54C4fZzQb9IX4I-5:-KKFYimx9b4FkxuqPN6i:110658269) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/server/context.go#L305)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Add a TODO to remove this after a short testing period? Removing both the reservation system and the old backoff code doesn't sound like a stable configuration.
</details>
Done.  And you're right, having neither is not stable.

---

*[storage/allocator_test.go, line 176 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKBzvD0Au2lCwXZ1f2W:-KKFGcyXpZkMGZe6dpBO:-470753674) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/allocator_test.go#L176)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Do these tests fail with reservations enabled? Since this is going to be the default we should generally be going through this code path in tests. Maybe we need to mock out the RPC itself (and that can be left as a TODO), but we should try to turn this on wherever we can (especially in `multiTestContext`)
</details>
I've turned it on for a number of tests.  There are only a few that fail (notably client_test).  I've added #7226 to deal with the ones that are still set to false.

---

*[storage/replica_command.go, line 2550 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKA_VIV-nWVuM9BMjc7:-KKAfjeUPldpPTfR0cRu:928395679) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/replica_command.go#L2550)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> The second sentence is superfluous, because it's implied from the first.
</details>
Agreed, removed.

---

*[storage/replica_command.go, line 2560 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC0PyDk9P-1hUcBOQ4:-KKFMHxsRRy_T7YP_58x:-48706055) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/replica_command.go#L2560)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Nit: We should be getting the NodeID and StoreID in the same way here. Maybe just pass `r.store.Ident` into `reserve`?
> 
> Why does `reserve` need the node and store ID for the sender but only the store ID for the target?
</details>
Oh, I like the idea of using r.store.Ident.  Fixed.

We only need the store ID because we have the node ID for that store already locally.  Technically, we have both the node IDs locally, but passing it in allows up to not have to lookup more than one store in the map. 

---

*[storage/store_pool.go, line 101 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC1I5rxS7wy9vv2rng:-KKFNoJWIdGLXT3b58Wi:1832366427) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L101)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It's weird that a throttled target is "alive" instead of either "dead" or a distinct value. In what situation would you want to treat these two cases the same?
</details>
We want to return the number of alive stores, regardless of them being throttled, so that there is no warning about not having enough store to replicate to.  

---

*[storage/store_pool.go, line 390 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC2Al-9DrBcPequV0o:-KKFVIutA2oUSucYfYH0:463090474) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L390)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> A throttled store should only be counted in `aliveStoreCount` if it matches the required attributes. 
</details>
I'm keeping the same style as the original.  Which is all non-dead stores are part of alive count, but only matched stores are returned in the storelist. 

---

*[storage/store_pool.go, line 426 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC2vctz8C9hfM0sRG_:-KKFVfuxjeKE3g5imUDr:2104061342) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L426)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/nodeID/toNodeID/
</details>
I've reverted the change of having the rlock followed by the lock, so this temp variable is no longer needed.

---

*[storage/store_pool.go, line 445 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKAaHOmNtFYwFPo6O0U:-KKFZe2wxs64q5FvOKJo:-1365503924) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L445)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Is there a timeout for this RPC? What happens to replicate queue performance if there's a node that's slow to respond to `Reserve` requests?
</details>
We can adjust the values as needed.  I'll put in a todo to investigate adding a very short timeout onto this call.

---

*[storage/store_pool_test.go, line 451 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC3ZhW0xPSXM0fHmFK:-KKFYdF8nMPtF-c7LXTy:-105626503) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool_test.go#L451)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Can you do `stopper.AddCloser(ln)` here so the caller doesn't have to?
</details>
Done, but kinda ugly.  :)

---

*[storage/store_pool_test.go, line 507 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC4-WE4n9LhnpKNA9y:-KKFWFftZ9hLKgS2loFF:-1054325829) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool_test.go#L507)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Space after `//`. Or just remove the comment; doesn't look like the regex usage matters.
</details>
removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
All comments addressed.
Getting failures in the acceptance tests.  Looking into them.

---

Review status: 1 of 13 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-:-KKFjVQ1MDC5B2Ve7LR9:-1319218385)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 12 of 12 files at r7, 11 of 11 files at r8.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/store_pool_test.go, line 451 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC3ZhW0xPSXM0fHmFK:-KKFlHLiXWIYaEKqQv2X:-296688728) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool_test.go#L451)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done, but kinda ugly.  :)
</details>
Now that you're doing that, you should probably return the listener's address instead of the listener itself.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 12 of 13 files reviewed at latest revision, 10 unresolved discussions.

---

*[storage/store_pool_test.go, line 451 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKC3ZhW0xPSXM0fHmFK:-KKFoXLcbEBnudhkih0r:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool_test.go#L451)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Now that you're doing that, you should probably return the listener's address instead of the listener itself.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/store_pool.go, line 445 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKAaHOmNtFYwFPo6O0U:-KKJBMY-R1UheAmO9Nsh:965363369) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L445)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> We can adjust the values as needed.  I'll put in a todo to investigate adding a very short timeout onto this call.
</details>
We need *some* timeout here so the replication queue doesn't get blocked for the duration of a network partition, and it's easy to add to the context. 

---

*[storage/store_pool.go, line 446 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKJCMmvp1rqmhtWiyjL:-KKJCMmvp1rqmhtWiyjM:123426441) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bfda785d5dbff324ea9734dbd901b1de6af4cf8/storage/store_pool.go#L446)):*
> ```Go
> 	// If a reservation is declined, be it due to an error or because it was
> 	// rejected, we mark the store detail as having been rejected so it won't
> 	// be considered for other checks until after timeoutForDeclinedReservations
> ```

Errors and rejections are different. If we get rejected, we can probably try again soon, but if there's an error the store may be down for a long time. (and retrying after a rejection is cheap; retrying after an error may have to wait for a timeout). Don't store the time at which the reservation was declined; store the time at which it is eligible for consideration again.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 14 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/store_pool.go, line 445 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKAaHOmNtFYwFPo6O0U:-KKK_yDQqHibe94Usqgq:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L445)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We need *some* timeout here so the replication queue doesn't get blocked for the duration of a network partition, and it's easy to add to the context. 
</details>
Done

---

*[storage/store_pool.go, line 446 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKJCMmvp1rqmhtWiyjL:-KKJeXRcI_N4H59oPb9A:-1038666939) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bfda785d5dbff324ea9734dbd901b1de6af4cf8/storage/store_pool.go#L446)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Errors and rejections are different. If we get rejected, we can probably try again soon, but if there's an error the store may be down for a long time. (and retrying after a rejection is cheap; retrying after an error may have to wait for a timeout). Don't store the time at which the reservation was declined; store the time at which it is eligible for consideration again.
</details>
I agree entirely, it's actually on my list of things to do once this is merged.  I just didn't want to change it for this PR.  I can if you prefer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I think I've fixed the acceptance tests and all comments.  So this should be ready to merge if I can get an LGTM.

---

Review status: 2 of 14 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-:-KKKa0twZj7bwv4StL-8:1435039687)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 14 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store_pool.go, line 445 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKAaHOmNtFYwFPo6O0U:-KKKp76ZuLn5GDC3Z_Ej:-1685323959) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L445)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done
</details>
1 second may cause issues with very slow links. However, in the interest of iterating on this, we can cross that bridge when we get there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-:-KKKpR7zoYl6gJuE_zmW:1417088721)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 14 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store_pool.go, line 445 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKAaHOmNtFYwFPo6O0U:-KKKqhXfi1cmtLIXl_Fn:-1755113446) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L445)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> 1 second may cause issues with very slow links. However, in the interest of iterating on this, we can cross that bridge when we get there.
</details>
I think we can set it as max offset * 2 or something like that in the end.  Want it very quick, but if it's too small, we'll be in trouble.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 2 of 14 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store_pool.go, line 445 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKAaHOmNtFYwFPo6O0U:-KKKr0UO9tPp1dL3yXP5:1589591584) ([raw file](https://github.com/cockroachdb/cockroach/blob/53b59367e5b78fd5985b9738355e840f77fbb853/storage/store_pool.go#L445)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I think we can set it as max offset * 2 or something like that in the end.  Want it very quick, but if it's too small, we'll be in trouble.
</details>
It's more like worst-case round-trip time plus some amount of processing time on the receiving end. Alternatively, making the reservation process async would work around this (not sure how hard that is to do).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

petermattis commented on storage/store_pool.go:
```diff
@@ -39,15 +42,24 @@ const (
+
+	// timeoutForDeclinedReservations is the amount of time to not consider the
+	// store available for up-replication after a reservation was declined.
+	timeoutForDeclinedReservations = 10 * time.Second
```
Can we set this to 1 second? Or lower? I know this hasn't been tuned, but I'm not clear what benefit it is providing. Reservation RPCs will not be overloading the receiving node in any current setup. But this setting will add another layer of throttling to rebalancing which is potentially harmful to allocation throughput.

BramGruneir replied:
Yep.  I'm splitting it into two values.  10 seconds for failures, 1 second for denials.

PR is forthcoming.

tamird commented:
Reviewed 12 of 12 files at r11.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2421 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKLIaJhcHj7daqHlWd6:-KKLIUhkcMBzv6OWPA9E:-1911898627) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca109de3d5dd5a30270a14fd9cd175bdf29b7ed/storage/store.go#L2421)):*
> ```Go
> func (s *Store) Reserve(req roachpb.ReservationRequest) roachpb.ReservationResponse {
> 	if s.metrics.capacity.Value() == 0 {
> 		// On startup, it takes some time before compute metrics is run. When
> ```

inconsistent "compute metrics" vs "ComputeMetrics"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
*[storage/store.go, line 2421 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147#-KKLIaJhcHj7daqHlWd6:-KKOobHorfeSewk88wna:-443195919) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca109de3d5dd5a30270a14fd9cd175bdf29b7ed/storage/store.go#L2421)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> inconsistent "compute metrics" vs "ComputeMetrics"
</details>
Fixing in #7255

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7147)*
<!-- Sent from Reviewable.io -->
